lib/src/common/cache/cache.dart

```diff
@@ -0,0 +1,300 @@
+
+  /// A stream of [CacheContext]s that dispatches when an item is released from
+  /// the cache.
+  ///
+  /// This does not necessarily mean that the item has been removed, only that it
+  /// has been marked as eligible for removal. The [didRemove] stream contains
+  /// information about true removal.
+  Stream<CacheContext<TIdentifier, TValue>> get didRelease =>
+      _didReleaseController.stream;
+
+  /// A stream of [CacheContext]s that dispatches when an item is removed from
+  /// the cache.
+  Stream<CacheContext<TIdentifier, TValue>> get didRemove =>
+      _didRemoveController.stream;
+
+  /// The stream of [CacheContext]s that dispatches when an item is updated in
+  /// the cache.
+  Stream<CacheContext<TIdentifier, TValue>> get didUpdate =>
+      _didUpdateController.stream;
+
+  /// Returns a value from the cache for a given [TIdentifier]. If the
+  /// [TIdentifier] is not present in the cache the value returned by the given
+  /// [valueFactory] is added to the cache and returned.
+  ///
+  /// All calls to [get] await a call to the current caching strategy's
+  /// [CachingStrategy.onDidGet] lifecycle method before returning the cached value.
+  /// If the [Cache] does not contain an instance for the given [TIdentifier],
+  /// the given [valueFactory] is called and a [CacheContext] event is emitted on
+  /// the [didUpdate] stream. A call to [get] that returns a cached value does
+  /// not emit this event as the [Cache] has not changed.
+  ///
+  /// Calls to [get] are evaluated synchronously and will return the future value
+  /// established by the first call to [get]. This allows calls to [get] to
+  /// return the same value regardless of completion of asynchronous calls to the
+  /// current caching strategy's [CachingStrategy.onDidGet] lifecycle method or
+  /// [valueFactory].
+  ///
+  ///     var a = cache.get('id', _superLongCall);
+  ///     var b = cache.get('id', _superLongCall);
+  ///     var c = cache.get('id', _superLongCall);
+  ///
+  ///     // The futures in this list will all resolve to the same instance,
+  ///     // which was returned by `_superLongCall` the first time `get` was called.
+  ///     var values = Future.wait([a, b, c]);
+  ///
+  /// If the [Cache] [isOrWillBeDisposed] then a [StateError] is thrown.
+  @mustCallSuper
+  Future<TValue> get(TIdentifier id, Func<TValue> valueFactory) {
+    _throwWhenDisposed('get');
+    _cachingStrategy.onWillGet(id);
+    // Await any pending cached futures
+    if (_cache.containsKey(id)) {
+      return _cache[id].then((TValue value) async {
+        await _cachingStrategy.onDidGet(id, value);
+        return value;
+      });
+    }
+
+    // Install Future value
+    final Completer<TValue> completer = new Completer<TValue>();
+    _cache[id] = completer.future;
+    try {
+      final TValue value = valueFactory.call();
+
+      _cachingStrategy.onDidGet(id, value).then((Null _) {
+        _didUpdateController.add(new CacheContext(id, value));
```

Comment:
In this case are we really adding the value to the cache? i.e. should this event be didAdd instead of didUpdate?